src/emcee.jl

```diff
@@ -0,0 +1,227 @@
+struct Ensemble{D} <: MHSampler
+    n_walkers::Int
+    proposal::D
+end
+
+struct Walker{T<:Union{Vector,Real,NamedTuple},L<:Real}
+    params::T
+    lp::L
+end
+
+Walker(t::Transition) = Walker(t.params, t.lp)
+Walker(model::DensityModel, params) = Walker(params, logdensity(model, params))
+
+# Define the first step! function, which is called at the 
+# beginning of sampling. Return the initial parameter used
+# to define the sampler.
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::Ensemble,
+    N::Integer,
+    ::Nothing;
+    init_params = nothing,
+    kwargs...,
+)
+    if init_params === nothing
+        return propose(spl, model)
+    else
+        return Transition(model, init_params)
+    end
+end
+
+# Define the other step functions. Returns a Transition containing
+# either a new proposal (if accepted) or the previous proposal 
+# (if not accepted).
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::Ensemble,
+    ::Integer,
+    params_prev;
+    kwargs...,
+)
+    # Generate a new proposal. Accept/reject happens at proposal level.
+    return propose(spl, model, params_prev)
+end
+
+#
+# Initial proposal
+# 
+function propose(spl::Ensemble, model::DensityModel)
+    # Make the first proposal with a static draw from the prior.
+    static_prop = StaticProposal(spl.proposal.proposal)
+    mh_spl = MetropolisHastings(static_prop)
+    return map(x -> Walker(propose(mh_spl, model)), 1:spl.n_walkers)
+end
+
+
+
+#
+# Every other proposal
+# 
+function propose(spl::Ensemble, model::DensityModel, walkers::Vector{W}) where {W<:Walker}
+    new_walkers = Vector{W}(undef, spl.n_walkers)
+    interval = 1:spl.n_walkers
+
+    for i in interval
+        walker = walkers[i]
+        other_walker = rand(walkers[interval.!=i])
```

Comment:
I guess it is more efficient to sample the index of the other walker instead of computing the subset of all other walkers and then sampling from them. I.e., something like
```suggestion
    nwalkers = spl.n_walkers
    others = 1:(nwalkers - 1)

    for i in interval
        walker = walkers[i]
        idx = mod1(i + rand(rng, others), nwalkers)
        other_walker = walkers[idx]
```

Reply:
That's a clever fix! I had been trying to get something like this but settled for the ugly solution.